### ISSUE TITLE
Packaging external node_modules

### ISSUE THREAD

**Issue by azaharakis** [mrgrain/cdk-esbuild#162]
**Is your feature request related to a problem? Please describe.**

I have a use case where an npm module I depend on uses readFileSync to pull in a file. This results in my bundle failing when executing in a cloudwatch synthetics context. 

> ERROR: Canary error: 
Error: ENOENT: no such file or directory, open '/opt/nodejs/dom-testing-library.js'

**Describe the solution you'd like**
Is there a way to configure the esbuild through this module to copy the node_modules, and have the bundle resolve the node_modules?

**Describe alternatives you've considered**
Rewriting the dependency so it does not do this...

**Additional context**
The dependency is https://github.com/testing-library/pptr-testing-library/blob/main/lib/index.ts#L8-L11

**Comment by azaharakis** [comment]
It looks like the `copyDir` prop might be what I want.
But using it say with `./node_modules/pptr-testing-library` will copy the contents into the root directory of the asset bundle ignoring the outDir: `nodejs/node_modules` that i'm using to bundle for `Synthetics` 

I changed this line
https://github.com/mrgrain/cdk-esbuild/blob/163b38c3940bdf4d796520dab5a2817dbc810161/src/bundler.ts#L121
to:
```
path.join(outputDir, 'nodejs/node_modules')
```

Thinking that by getting the assets in the same location of the bundle the intent would be to package it and send it to s3.
However the error I get when I use the changes looks like:

```
Bundling asset FooBar0/TypeScriptCode/Stage...
Error: Failed to bundle asset FooBar0/TypeScriptCode/Stage, bundle output is located at /Users/azaharakis/FooBar/cdk.out/bundling-temp-48a6bfb0079015bedd2abb1e6747dd6ea88fcf0d4e38044a5e19506e68766d7b-error: TypeError: messages is not iterable
```

Not entirely familiar with es-build and synthetics packaging requirements. So still poking around the logs of the output

**Comment by mrgrain** [comment]
Apologies, I've been travelling. Will look at this asap.

I think the issue is that esbuild is bundling all dependencies into a single file and pptr-testing-library then gets confused about where the files should be. There should be a way to tell esbuild how to do this. The easiest thing to do might be to just NOT bundle.

**Comment by azaharakis** [comment]
Hi @mrgrain, no apologies needed, wasn't even expecting a response tbh :)

> I think the issue is that esbuild is bundling all dependencies into a single file and pptr-testing-library then gets confused about where the files should be

Yep thats right. pptr-testing-library is unable to resolve the dependency because readFileSync isn't evaluated by esbuild or any bundler for that matter the same way as `import`/`require` is.

>  The easiest thing to do might be to just NOT bundle.

I would have thought so as well. But i've had difficult finding good examples of synthetics asset implementations that take this approach on anything non trivial. From what I understand i'd need to:
Upload my asset as a zip to s3 in the directory structure that looks something like this:
```
nodejs/
├─ node_modules/
│  ├─ pptr-testing-library
│  ...
│  ├─ canary-handler.js
```
But there are a couple of issues i've found so far going down this approach.

1. This [construct](https://docs.aws.amazon.com/cdk/api/v1/docs/aws-s3-deployment-readme.html) allows you to statically define what you want to be uploaded to s3 (Ideally I wanted to avoid writing anything custom or any procedural scripts to do this), but it unzips the contents in the bucket instead of keeping it zipped e.g https://github.com/aws/aws-cdk/issues/8065. This is incompatible with Synthetics, Synthetics wants your asset zipped in the bucket.

2. If you do manage to upload a zip to s3. To make changes to the canary seems to at least require changing the object key e.g `my-bucket/canary.asset{timestamp}.zip` in order to trigger an update to the stack so synthetics can rebuild the layer from a new zip. If the file name stays the same, but the zip contents change the layer is not updated.

These were the 2 issues prevented me from going down this path further and was hopeful that this library was abstracting the upload to s3 functionality as part of its implementation that would handle all cases for asset publishing to synthetics.

**Comment by mrgrain** [comment]
Good questions. This package is using S3Asset to upload the zip to S3. I just have a tiny wrapper around it.
https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_assets-readme.html


Initially I thought it would be enough to do
```
buildOptions {
  bundle: false
}
```

but I recall now that Synthetics has some odd requirements with the file system.

**Comment by mrgrain** [comment]
> Upload my asset as a zip to s3 in the directory structure that looks something like this:

Agreed, that should be the goal and it should be possible with the right combination of outdir, copydir and external.

We also might need an extra `node_modules` in the tree structure, but that should be easy enough if needed.

PS: I'm aware you probably know all this already. I'm really just writing down my own thought process so I can come back to it later.

**Comment by mrgrain** [comment]
tl;Dr This is currentlynot easily possibly with this package or any esbuild features.



I've played around with this on a separate branch. This is the best I came up with. It's untested, but creates the files in the right place:
https://github.com/mrgrain/cdk-esbuild/blob/pptr-testing-library/examples/website/infrastructure/stack.ts#L46

The current implementation of Copydir isn't clever enough to copy the dependency to the correct dir. It's a nice feature idea though and should be easy enough to add support for a map of multiple things to copy into various locations. I don't just want to change it to copy to the outdir, because I originally added the feature to be able to copy files NOT to the outdir.

**Comment by azaharakis** [comment]
@mrgrain thank you so much for pointing me in the direction to un-block myself with your provided branch! This particular issue was outside of the scope of this package imo so I really appreciate your help on this!!

> The current implementation of Copydir isn't clever enough to copy the dependency to the correct dir. It's a nice feature idea though and should be easy enough to add support for a map of multiple things to copy into various locations. I don't just want to change it to copy to the outdir, because I originally added the feature to be able to copy files NOT to the outdir.

So I understand you are suggesting:
```
type copyDir: string
````
extended to say:
```
type copyDir: string | Record<string,string>
```
Where in the case of a map the id represents the `srcDir` and the value representing the `destinationDir` relative to projects `src`, not the `outDir` or if provided from the relative path of `absWorkingDir `?

**Comment by mrgrain** [comment]
> extended to say:
> 
> `type copyDir: string | Record<string,string>`

Yes, that's right. We could also add a string array for even more options:

```
type copyDir: string | string[] | Record<string,string>
```

> Where in the case of a map the id represents the `srcDir` and the value representing the `destinationDir` relative to projects `src`, not the `outDir` or if provided from the relative path of `absWorkingDir`?

We need to be a bit careful here to not mix up the various things. `copyDir` is a feature added by `cdk-esbuild` (this package) and has nothing to do with `esbuild`. [I did suggest to add it to esbuild directly](https://github.com/evanw/esbuild/issues/982), but they understandably didn't want to implement it.

For CDK we need to copy the files into an asset staging area. That's one of the `cdk.out/asset.1234hash` directories you can see. Let's call this `assetDir`.

I've also noticed that `esbuild` uses the id of the map as target, and the valuer as source. It would probably be a good idea to follow that pattern for consistency.


Putting all things together, in the case of a map the id represents the `destinationDir` relative to the `assetDir` and the value represents the `srcDir` relative to the current working dir (same logic as currently done for the string).

These would then all be equivalent:

```
copyDir: "path/to/files"
copyDir: ["path/to/files"]
copyDir: {
  ".": "path/to/files"
}
```

PS: I can't remember why I called it `copyDir`. Pretty sure it can also copy a single file. 😅

**Comment by azaharakis** [comment]
Hi @mrgrain, nice one! The implementation looks great! i'll need a bit of time to test it for my use case. Will try to get to it before Monday

**Comment by mrgrain** [comment]
@azaharakis I'm probably going to release this on the weekend. No worries if you don't get the testing it. We can always iterate afterwards!

**Comment by azaharakis** [comment]
@mrgrain so sorry mgrain, i've been super swamped and I agree and think you should move forward, the implementation looks solid to me

**Comment by mrgrain** [comment]
@azaharakis v3.4.0 has been released. I've updated the example branch as well:
https://github.com/mrgrain/cdk-esbuild/blob/a6ce259134419f2858a749bc60131b226ee9affe/examples/website/infrastructure/stack.ts#L46-L59

Please feel free to comment/re-open/open a new issue if this doesn't get you all the way. =D
